contrib/platform/src/com/sun/jna/platform/win32/WinCrypt.java

```diff
@@ -72,10 +72,16 @@ public DATA_BLOB(Pointer memory) {
         }
 
         public DATA_BLOB(byte [] data) {
-            pbData = new Memory(data.length);
-            pbData.write(0, data, 0, data.length);
+            super();
+            if (data.length > 0) {
+                pbData = new Memory(data.length);
+                pbData.write(0, data, 0, data.length);
+            } else {
+                // For zero length array we just need a pointer to any allocated memory, the
+                // data there will be ignored
+                pbData = getPointer();
```

Comment:
This will become problematic when you call `Pointer#clear` on `pbData`. If then `Structure#read` is invoked on the `DATA_BLOB` structure, the content changes, although nothing "did" that. While I don't see an immediate problem, I learned, that side effects come back to haunt us.

I think the `Memory(1)` solution is ugly, but i don't see an obvious performance problem. If the hot path involves creating `DATA_BLOB`s from empty arrays, the issue is somewhere else.

Reply:
Fair enough, although the clearing specifies 0 length so it shouldn't impact.  Still, a fresh new pointer avoids any potential issues.

Reply:
If you believe that `Memory(1)` is ugly, how about to allocate memory by `LocalAlloc` instead of `malloc`? And then free it explicitly within `CryptProtectData` and `CryptUnprotectData` using `LocalFree` (as we do with the result). `LocalAlloc` behaves nicely with size=0 in contrast to `malloc`.
But in my opinion, solution with `Memory(1)` is quite safe and reliable and assumes minimal code changes. However it appeared it isn't very obvious.

Reply:
I think `Memory(1)` is the best solution here (together with a comment why a 1 byte memory area is allocated, although a 0 lenght byte array is supplied). Ugly in this case means, that I think MS should not rely on definition hole in the C specification.

Reply:
This is not a suggestion to actually do this, but I did go down the road of thinking about `new Pointer(Native.malloc(0));` (and eventually freeing it).  Just to show there's an uglier solution than `Memory(1)`. :)